docs/source/advanced/contrib_thunder.rst

```diff
@@ -0,0 +1,468 @@
+On one hand, fewer primitive operations can make program transformation, and execution easier.
+Fewer primitives means fewer transformation rules – since transformation rules are defined on primitives – and a smaller interface with executors.
+On the other hand, too few primitive operations may make it hard, or impossible, to express all the operations that practitioners are interested in.
+Too few primitive operations may also make it difficult to execute programs quickly and numerically accurately.
+
+For example, the ``expm1`` operation can mathematically be defined in terms of the ``exp`` and subtraction operations,
+and so it does not need to be a primitive to enable any functionality.
+Many libraries, including `C++’s standard library <https://en.cppreference.com/w/c/numeric/math/expm1>`_,
+still define an ``expm1`` operation for numerical accuracy, and so ``expm1`` is a
+`primitive <https://github.com/Lightning-AI/lightning-thunder/blob/888b46324462fba70f93d5017bc0d99025f05091/thunder/core/prims.py#L1791>`_ in *Thunder*.
+
+
+-----------------
+The Core Language
+-----------------
+Above the primitives is the ``core`` language, or `clang <https://github.com/Lightning-AI/lightning-thunder/blob/main/thunder/clang/__init__.py>`_.
+Clang operations are mostly written like any other Python operation.
+They ultimately call the primitive operations, although they may call other operations before doing so
+(for example, ``clang.foo`` might call ``clang.bar`` which calls ``prims.bar``).
+
+Core language operations are intended to be common functionality that’s useful when
+defining practitioner-facing language like ``torch`` or ``numpy``.
+Many of these operations are just wrappers around primitive operations.
+For example, the elementwise binary primitives are as simple as possible, so they don’t perform broadcasting or type promotion.
+The core language elementwise binary operations, however, do perform broadcasting and type promotion.
+
+Before adding a core language operation consider if the functionality expressed is universal enough.
+
+As a style note, operations in *Thunder* should defer as much error checking as possible.
+For example, if a primitive’s meta function will perform an error check for ``X``,
+then the core language operation that calls it should generally not also check for ``X``.
+
+
+------------------
+The Torch Language
+------------------
+To translate ``torch`` operations into something that *Thunder* understands we define a
+`torch language <https://github.com/Lightning-AI/lightning-thunder/blob/main/thunder/torch/__init__.py>`_.
+Operations in the ``torch`` language should be written as if a practitioner were writing them,
+and they should reflect the behavior of their corresponding torch operations (small deviations are sometimes OK).
+
+When a program is interpreted, torch operations are remapped into these operations, which ultimately call primitive operations.
```

Comment:
I would expand this section a bit, also adding or linking to an easy example that shows how a torchlang op matches a pytorch operator and how it calls different prims etc. Especially since this is the most popular language for our users.

Reply:
Or hint at the example below (I hadn't reached it when I posted the above) for those who want to see it already

Reply:
I will update outline the structure and how it is supposed to be read.